src/pudl/transform/ferc1.py

```diff
@@ -242,31 +242,42 @@ def drop_duplicate_rows_dbf(df, params: DropDuplicateRowsDbf):
     duplicates are dropped with ``keep="first"``. If any duplicates do not contain the
     same data, an assertion will be raised.
     """
-    if params is None:
-        params = DropDuplicateRowsDbf()
-    if params.data_columns:
-        pks = (
-            pudl.metadata.classes.Package.from_resource_ids()
-            .get_resource(params.table_name.value)
-            .schema.primary_key
-        )
-        # checks to make sure the drop is targeted as expected
-        pk_dupes = df[df.duplicated(pks, keep=False)]
-        pk_data_dupes = pk_dupes[
-            pk_dupes.duplicated(pks + params.data_columns, keep=False)
-        ]
-        dupes_w_unique_data = pk_dupes.loc[
-            pk_dupes.index.difference(pk_data_dupes.index)
-        ]
-        if not dupes_w_unique_data.empty:
-            raise AssertionError(
-                "Duplicates have unique data and should not be dropped. Unique data: "
-                f"{len(dupes_w_unique_data)}: {dupes_w_unique_data}"
+    pks = (
+        pudl.metadata.classes.Package.from_resource_ids()
+        .get_resource(params.table_name.value)
+        .schema.primary_key
+    )
+    # add a column that indicates whether or not any of the data columns contain null data
+    df.loc[:, "null_data"] = df[params.data_columns].isnull().any(axis="columns")
+
+    # checks to make sure the drop is targeted as expected
+    # of the PK dupes, drop all instances when the data *is also the same*
+    dupes_w_unique_data = (
+        df[df.duplicated(pks, keep=False)].drop_duplicates(
+            pks + params.data_columns, keep=False
+        )
+        # if there are pk+data dupes, is there one record with some null data
+        # an other with completely non-null data??
+        .assign(
+            count_of_null_vs_non_null=lambda x: x.groupby(pks)[["null_data"]].transform(
+                "nunique"
             )
-        logger.info(
-            f"Dropping {(len(pk_data_dupes)/2)/len(df):.1%} of duplicate records."
         )
-        df = df.drop_duplicates(pks, keep="first")
+    )
+    if not dupes_w_unique_data[
+        dupes_w_unique_data.count_of_null_vs_non_null != 2
+    ].empty:
+        raise AssertionError(
+            "Duplicates have unique data and should not be dropped. Unique data: "
+            f"{len(dupes_w_unique_data)}: {dupes_w_unique_data}"
+        )
+    len_og = len(df)
+    df = (
+        df.sort_values(by=["null_data"], ascending=True)
+        .drop_duplicates(pks, keep="first")
+        .drop(columns=["null_data"])
+    )
```

Comment:
This seems like a good way to make sure to keep the non null dups (as long as ascending=True puts False above True!)

Reply:
it does! I always have to double check and I did add a case into the unit tests to ensure this